### PR TITLE
prov/tcp: Adding completion semantic & tx/rx path optimizations

### PR DESCRIPTION
**Pull request by vkrishna**

### CODE REVIEW

File: prov/tcp/src/tcpx.h [hunk 1]
```diff
@@ -141,11 +141,18 @@ struct tcpx_rx_detect {
 
 typedef void (*tcpx_ep_progress_func_t)(struct tcpx_ep *ep);
 
+typedef int (*tcpx_rx_process_fn_t)(struct tcpx_xfer_entry *rx_entry);
+
+struct tcpx_cur_rx {
+	struct tcpx_xfer_entry	*rx_entry;
+	tcpx_rx_process_fn_t 	process_fn;
+};
```
Comment (shefty) [inline prov/tcp/src/tcpx.h]: Is there a reason to create a sub-structure, versus embedding the 2 fields directly

File: prov/tcp/src/tcpx_cq.c [hunk 2]
```diff
@@ -86,9 +86,10 @@ struct tcpx_xfer_entry *tcpx_xfer_entry_alloc(struct tcpx_cq *tcpx_cq,
 void tcpx_xfer_entry_release(struct tcpx_cq *tcpx_cq,
 			     struct tcpx_xfer_entry *xfer_entry)
 {
-	if (xfer_entry->ep->cur_rx_entry == xfer_entry)
-		xfer_entry->ep->cur_rx_entry = NULL;
-
+	if (xfer_entry->ep->cur_rx.rx_entry == xfer_entry) {
+		xfer_entry->ep->cur_rx.rx_entry = NULL;
+		xfer_entry->ep->cur_rx.process_fn = NULL;
```
Comment (shefty) [inline prov/tcp/src/tcpx_cq.c]: Can only one of these bet cleared?

File: prov/tcp/src/tcpx_progress.c [hunk 3]
```diff
@@ -217,6 +219,9 @@ static void tcpx_prepare_rx_remote_read_resp(struct tcpx_xfer_entry *resp_entry)
 	resp_entry->done_len = 0;
 
 	tcpx_tx_queue_insert(resp_entry->ep, resp_entry);
+	resp_entry->ep->cur_rx.rx_entry = NULL;
+	resp_entry->ep->cur_rx.process_fn = NULL;
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c]: Can only one  of these be cleared?

File: prov/tcp/src/tcpx_progress.c [hunk 4]
```diff
@@ -341,22 +349,23 @@ static int tcpx_get_rx_entry(struct tcpx_rx_detect *rx_detect,
 		rx_entry->msg_hdr = rx_detect->hdr;
 		rx_entry->msg_hdr.hdr.op_data = TCPX_OP_READ_RSP;
 		rx_entry->done_len = sizeof(rx_detect->hdr);
+		new_rx->process_fn = process_rx_read_entry;
 		break;
 	default:
 		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,
 			"invalid op in hdr\n");
 		return -FI_EINVAL;
 	}
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c]: You might be able to create an array of function pointers, assign those to the ep, and use the opcode as an index into the array (after validating it).  That would avoid the switch statement completely.

**Comment by shefty** [comment]
This PR causes hangs in OS X.   (I tried re-starting the CI already.)

**Comment by vkrishna** [comment]
I will check and update the PR.

**Comment by shefty** [comment]
Hmm... then the hang is more reproducible - since it happened 4 times in a row. @a-ilango has there been a recent change in rxm in this area that hit master?

**Comment by a-ilango** [comment]
>  has there been a recent change in rxm in this area that hit master?
There isn't any recent change that went into the CM thread area.

**Comment by vkrishna** [comment]
Yes it is now consistent for me locally as well. 

I checked out Arun's PR #4391 and cherry-picked these commits on top of it and checked locally. I can't reproduce the failure anymore.

**Pull request merged**